internal/core/src/segcore/DeletedRecord.h

```diff
@@ -17,139 +17,337 @@
+        : insert_record_(insert_record),
+          segment_(nullptr),
+          deleted_lists_(SortedDeleteList::createInstance()) {
     }
 
-    std::shared_ptr<TmpBitmap>
-    clone_lru_entry(int64_t insert_barrier,
-                    int64_t del_barrier,
-                    int64_t& old_del_barrier,
-                    bool& hit_cache) {
-        std::shared_lock lck(shared_mutex_);
-        auto res = lru_->clone(insert_barrier);
-        old_del_barrier = lru_->del_barrier;
-
-        if (lru_->bitmap_ptr->size() == insert_barrier &&
-            lru_->del_barrier == del_barrier) {
-            hit_cache = true;
-        } else {
-            res->del_barrier = del_barrier;
+    ~DeletedRecord() {
+    }
+
+    DeletedRecord(DeletedRecord<is_sealed>&& delete_record) = delete;
+
+    DeletedRecord<is_sealed>&
+    operator=(DeletedRecord<is_sealed>&& delete_record) = delete;
+
+    void
+    LoadPush(const std::vector<PkType>& pks, const Timestamp* timestamps) {
+        if (pks.empty()) {
+            return;
+        }
+
+        auto max_deleted_ts = InternalPush(pks, timestamps);
+
+        if (max_deleted_ts > max_load_timestamp_) {
+            max_load_timestamp_ = max_deleted_ts;
         }
 
-        return res;
+        //TODO: add support for dump snapshot when load finished
     }
 
+    // stream push delete timestamps should be sorted outside of the interface
+    // considering concurrent query and push
     void
-    insert_lru_entry(std::shared_ptr<TmpBitmap> new_entry, bool force = false) {
-        std::lock_guard lck(shared_mutex_);
-        if (new_entry->del_barrier <= lru_->del_barrier) {
-            if (!force ||
-                new_entry->bitmap_ptr->size() <= lru_->bitmap_ptr->size()) {
-                // DO NOTHING
-                return;
+    StreamPush(const std::vector<PkType>& pks, const Timestamp* timestamps) {
+        if (pks.empty()) {
+            return;
+        }
+
+        InternalPush(pks, timestamps);
+
+        // disable dumpsnapshot function temporarily because corner case
+        // bool can_dump = timestamps[0] >= max_load_timestamp_;
+        // if (can_dump) {
+        //     DumpSnapshot();
+        // }
+    }
+
+    Timestamp
+    InternalPush(const std::vector<PkType>& pks, const Timestamp* timestamps) {
+        int64_t removed_num = 0;
+        int64_t mem_add = 0;
+        Timestamp max_timestamp = 0;
+
+        SortedDeleteList::Accessor accessor(deleted_lists_);
+        for (size_t i = 0; i < pks.size(); ++i) {
+            auto deleted_pk = pks[i];
+            auto deleted_ts = timestamps[i];
+            if (deleted_ts > max_timestamp) {
+                max_timestamp = deleted_ts;
+            }
+            std::vector<SegOffset> offsets;
+            if (segment_) {
+                offsets =
+                    std::move(segment_->search_pk(deleted_pk, deleted_ts));
+            } else {
+                // only for testing
+                offsets = std::move(
```

Comment:
Do not add test code here in the future.
You should encapsulate the `search_pk` as a function pointer within DeleteRecord. Pass the function object from outside.